### PR TITLE
feat(eslint): alphanumeric sorting

### PR DESCRIPTION
**Pull request by stevensacks**
Alphanumeric sorting of imports, PropTypes, Component props in JSX, and destructured props.

I'd like to talk about what I think is a very important change and uses a few rules to accomplish and that is alphanumeric sorting throughout.

Every developer has a different idea about what a "logical group" is. I used to group my props by what I thought made logical sense. However, even that thought process evolves, and I found myself looking back at old code and reordering props to match my more recent opinion about it. This created a lot of fragmentation over time.

There's one logical grouping that everyone knows instinctively, and that is alphanumeric. There's no debating it. It's not a matter of opinion that the letter B comes before the letter P, or that 4 is greater than 2. We don't have to think about it. It's automatic.

Alphanumerically sorting imports, props, destructured object properties, (and react hook dependency arrays) results in everything being super easy to find when looking at your and other people's code because it's sorted by a pattern that we all agree on without thinking. The great news is that with the exception of the import sorting rule, the sorting rules are all --fix-able.

### CODE REVIEW

**Comment by igetgames** [comment]
This one I could see being part of a `eslint-config-amex/strict` addon, with more opinionated rulesets.

It would be fun to watch editor autofix with these on.

**Comment by smackfu** [comment]
(Reupping my comment from the closed PR.)
I'm not opposed to this rule in theory, but I have concerns because:
a) it directly contradicts our previous import order rule (which was "builtin", "external", "internal")
b) there is no auto-fix for some reason (why?)

As a practical matter, I'm not sure what the path would be to getting rid of these warnings in every single existing file.

**Comment by stevensacks** [comment]
I just looked into it and it looks like the import/order eslint rule has autofix. I will configure it to behave the same and test it.

https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/order.md

As for your previous import order rule, I explained why that ordering is problematic in my post above, but to expand on it a little, I proffer this to you.

If you didn't know what the ordering rule was, and you looked at a file with a lot of imports, how long would it take you to figure out what the ordering was so that you could find the import that you were looking for? I mean, even if you knew what the ordering rule was, how long would it take you to figure it out?

Now, how long does it take you to think about whether B comes before K? Or that 4 is greater than 1? The answer is effectively zero. You don't have to learn and remember the sorting rule. It's your default.

**Comment by PixnBits** [comment]
>There's no debating it.

I'd like to be careful here. There's no debating the algorithm (well, even then, is it ASCII/UTF code sorted or character sorted? ex: a,à,b or a,b,à) but the main debate for using the rule is: is the algorithm _useful_ (to the maintainers of the code)?

The order someone chooses might be based off of dependencies being relevant to each other, if their names start with different letters then the logical association is broken. Currently we're using "builtin", "external", "internal" as @smackfu mentioned to ease identification of what should and should not be installed as a dependency (ex: check out the downloads of [`fs`](https://www.npmjs.com/package/fs)).

File: index.js [hunk 1]
```diff
@@ -138,6 +139,37 @@ module.exports = {
     // https://reactjs.org/docs/hooks-rules.html#eslint-plugin
     'react-hooks/rules-of-hooks': 'error',
 
+    // Alphabetical sorting is the most reliable way to make code consistently readable
```
Comment (PixnBits) [inline index.js:142]: [is it the most useful](https://github.com/americanexpress/eslint-config-amex/pull/20#issuecomment-621302651)? If not, what are the tradeoffs and why was reliability chosen over utility? (please add the answers, etc to the comment for future maintainers)

**Comment by stevensacks** [comment]
> > There's no debating it.
> 
> I'd like to be careful here. 

Let me be clear about this point.

There's no debating that human beings know that 4 is greater than 1. You don't have to think about it.

There's no debating that human beings who use the English alphabet know the order of letters in the alphabet A through Z. In order to write Javascript, you have to write in English even if your native tongue uses a different alphabet.

Any concerns about proper ordering of unicode characters with accents in the context of this rule are unfounded. Using extended unicode characters for variable names in Javascript is a no-no.

You do not have to think about alphanumeric. It's automatic.

As far as imports are concerned, you have to think about what is internal and external, but more importantly, how is that grouping relevant? I would argue that it's not. It really doesn't matter if the code you're importing comes from a node_modules export or one of your own. You needed that function so you imported it. You use it like any other Javascript, there's no special treatment. So grouping by that classification doesn't impart any value and is therefore arbitrary.

As far as ordering by relevancy, what does it meant to be relevant? It's circumstantial. And what's relevant to you while you're writing it may not be relevant to somebody else (or future you) when reading it.

This rule, for me, is about making code consistently readable and reducing unnecessary refactoring. The need to quickly find what you're looking for in your or other people's code. Logical relevance is a matter of opinion. Alphanumeric is not.

What seems logical right now might change tomorrow or a week from now. Rather than reorder things in the future based on what is logical at that time, alphanumeric means I don't need to worry about reordering in the future. It saves unnecessary refactoring, which makes code reviews and git diffs easier.

Believe me, I empathize with the desire to order by logical relevance. I still sometimes want to group like that. But, I know through experience in using this methodology that alphanumeric ordering applied consistently throughout code makes it much easier for myself and others to quickly be able to read code and find what I'm looking for and results in less "fiddling" with stuff that doesn't really matter.

**Pull request closed without merging**